hyperspy/_signals/lazy.py

```diff
@@ -521,87 +551,98 @@ def _map_all(self, function, inplace=True, **kwargs):
 
     def _map_iterate(self,
                      function,
-                     iterating_kwargs=(),
+                     iterating_kwargs={},
                      show_progressbar=None,
                      parallel=None,
                      max_workers=None,
-                     ragged=None,
+                     ragged=False,
                      inplace=True,
+                     output_signal_size=None,
+                     output_dtype=None,
                      **kwargs):
-        if ragged not in (True, False):
-            raise ValueError('"ragged" kwarg has to be bool for lazy signals')
-        _logger.debug("Entering '_map_iterate'")
-
-        size = max(1, self.axes_manager.navigation_size)
-        from hyperspy.misc.utils import (create_map_objects,
-                                         map_result_construction)
-        func, iterators = create_map_objects(function, size, iterating_kwargs,
-                                             **kwargs)
-        iterators = (self._iterate_signal(), ) + iterators
-        res_shape = self.axes_manager._navigation_shape_in_array
-        # no navigation
-        if not len(res_shape) and ragged:
-            res_shape = (1,)
-
-        all_delayed = [dd(func)(data) for data in zip(*iterators)]
-
-        if ragged:
-            if inplace:
-                raise ValueError("In place computation is not compatible with "
-                                  "ragged array for lazy signal.")
-            # Shape of the signal dimension will change for the each nav.
-            # index, which means we can't predict the shape and the dtype needs
-            # to be python object to support numpy ragged array
-            sig_shape = ()
-            sig_dtype = np.dtype('O')
-        else:
-            one_compute = all_delayed[0].compute()
-            # No signal dimension for scalar
-            if np.isscalar(one_compute):
-                sig_shape = ()
-                sig_dtype = type(one_compute)
+        # unpacking keyword arguments
+        nav_indexes = self.axes_manager.navigation_indices_in_array
+        if ragged and inplace:
+            raise ValueError("Ragged and inplace are not compatible with a lazy signal")
+        chunk_span = np.equal(self.data.chunksize, self.data.shape)
+        chunk_span = [chunk_span[i] for i in self.axes_manager.signal_indices_in_array]
```

Comment:
The ragged inplace stuff isn't really necessary... That is just a holdover from previous code where it wasn't possible.  I can replace that.

The line below that is very important.  The way the function works you have to have a chunk that spans the signal dimension.  We are basically passing the chunk to the wrapping `process_dask_chunk` function which then applies the function to the data.  If you didn't have a chunk that spanned the signal dimension then things would break anytime you tried to do anything with a signal.

Reply:
I wanted to just fix the problems here rather than make any larger changes which is why I kept the ragged and in place message.  That is how the code used to operate so I didn't want to break anything that might depend on it in a patch.

Reply:
OK, if I understand correctly, the current implementation of `process_dask_chunk` doesn't support chunking in the signal dimension and this could be an additional improvement which should be done in a separate PR since there are no reason why it shouldn't be possible. One for #2669.

Reply:
There are a couple of pretty big problems with chunking in the signal dimension.  For something simple like adding some value to every pixel in the dataset then there are no reasons why you couldn't chunk the signal dimension.  For something more complicated like finding a peak in a 2D signal or converting from carteisian coordinates to polar coordinates this is significantly more difficult.  For the second example we would need to use [dask.array.map_overlap](https://docs.dask.org/en/latest/array-overlap.html) which complicates things a fair amount.  

So while it is possible, and we might want to consider doing this, I can't think of too many signals for some dataset that are that large.  Even a 32 bit 4096 x 4096 image is only around 70 megabytes or so which is still on the small size for 1 chunk.  Even a computer with 4 Gb of ram could easily handle that.

The case where I am maybe more interested in is the case where you have a 4D dataset and you want to do something like find a peak in 4 dimensions so you do need to use the `dask.array.map_overlap` function and possibly apply a function to both the signal and navigation dimensions.  I think that should be a different map function entirely, which allows any chunking and operates on either the entire array or a subset of the dimensions.

Reply:
I agree with @CSSFrancis on this. Assuming the chunks span the whole signal dimensions greatly reduce the complexity, and potentially avoid a whole host of possible bad behaviors. The obvious ones are the ones he already mentioned, where you need the whole signal dimension (like a full diffraction pattern), to get the useful information out.

Most of the things people want to apply via `.map`, where the signal chunking can be kept, can be directly applied via `dask.array` functions, which typically retains the chunking.